### PR TITLE
Support exception handling for finally regions in SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to #5710 

```C#
try
{
  try
  {
    CanThrow()
  }
  finally
  {
    // This is processed in happy-path and exception-path
    // Happy-path continues to DoSomething
    // Exception-path continues to outer finally
  }
  DoSomething() // This is process only in happy-path
}
finally
{
    // This is processed in happy-path and exception-path
    // Happy-path continues to DoSomethingElse
    // Exception-path reaches exit
}
DoSomethingElse() // This is process only in happy-path
```

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 1]
```diff
@@ -47,6 +46,24 @@ public void Loops_InstructionVisitedMaxTwice(string loop)
                 .And.ContainSingle(x => x != null && x.HasConstraint(TestConstraint.First) && !x.HasConstraint(BoolConstraint.True));
         }
 
+        [TestMethod]
+        public void Loops_InstructionVisitedMaxTwice_ForEach()
+        {
+            var code = @"
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:52]: pedantic: can be a constant

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs [hunk 2]
```diff
@@ -107,16 +119,66 @@ public void Finally_Nested_InstructionAfterFinally()
...
+            validator.TagStates("InOuterFinally").Should().HaveCount(2)
+                .And.ContainSingle(x => x.Exception == null)
+                .And.ContainSingle(x => x.Exception != null);
+            validator.TagStates("AfterOuterFinally").Should().HaveCount(1)  // Not visited by flows with Exception
+                .And.ContainSingle(x => x.Exception == null);
+            validator.ValidateExitReachCount(2);
+        }
+
+        [TestMethod]
+        public void Finally_Nested_InstructionAfterFinally_NoThrowsInFinally()
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs:148]: Could you please add a test with try-finally in a finally block?

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
Looks good so far but it would be great to add a few more tests.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs:148]: Good catch, there was an issue with that scenario

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**